lib/galaxy/jobs/runners/util/cli/job/lsf.py

```diff
@@ -0,0 +1,102 @@
+# A simple CLI runner for slurm that can be used when running Galaxy from a
+# non-submit host and using a Slurm cluster.
+from logging import getLogger
+
+try:
+    from galaxy.model import Job
+    job_states = Job.states
+except ImportError:
+    # Not in Galaxy, map Galaxy job states to Pulsar ones.
+    from pulsar.util import enum
+    job_states = enum(RUNNING='running', OK='complete', QUEUED='queued', ERROR="failed")
+
+from ..job import BaseJobExec
+
+log = getLogger(__name__)
+
+argmap = {
+    'memory': '-M',
+    'cores': '-n',
+    'queue': '-q',
+    'working_dir': '-cwd'
+}
+
+
+class LSF(BaseJobExec):
+
+    def __init__(self, **params):
+        self.params = {}
+        for k, v in params.items():
+            self.params[k] = v
+
+    def job_script_kwargs(self, ofile, efile, job_name):
+        scriptargs = {'-o': ofile,
+                      '-e': efile,
+                      '-J': job_name}
+
+        # Map arguments using argmap.
+        for k, v in self.params.items():
+            if k == 'plugin':
+                continue
+            try:
+                if not k.startswith('-'):
+                    k = argmap[k]
+                if k == 'memory':
+                    # Memory requires both -m and -R rusage[mem=v] request
+                    scriptargs['-R'] = "\"rusage[mem=%s]\"" % v
+                scriptargs[k] = v
+            except Exception:
+                log.warning('Unrecognized long argument passed to LSF CLI plugin: %s' % k)
+
+        # Generated template.
+        template_scriptargs = ''
+        for k, v in scriptargs.items():
+            template_scriptargs += '#BSUB %s %s\n' % (k, v)
+        return dict(headers=template_scriptargs)
+
+    def submit(self, script_file):
+        return 'bsub <%s' % script_file
```

Comment:
Not sure if this is going to play nicely with Popen upstream though (I mean the internal redirect).

Reply:
https://www.ibm.com/support/knowledgecenter/en/SSETD4_9.1.2/lsf_command_ref/bsub.1.html 

Looks like if it is just the last argument to bsub you don't need to pipe it into stdin?

Reply:
If that is still a problem though - we can make submit return something more structured - we probably shouldn't be so relaxed with argument handling and such here anyway. Returning something with a list of command and arguments as well as options for stdin/stdout/etc... would be probably be good.

Reply:
No worries, this is all working like this, so I wouldn't give it any importance. On the other LSF runner implementation I tried not using stdin and it never worked. But this is working as it is, so no worries here.